presto-spark/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskCompiler.java

```diff
@@ -0,0 +1,340 @@
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterators;
+import io.airlift.slice.DynamicSliceOutput;
+import io.airlift.slice.SliceOutput;
+import io.airlift.slice.Slices;
+import io.airlift.units.DataSize;
+import org.apache.spark.util.CollectionAccumulator;
+import scala.Tuple2;
+
+import javax.inject.Inject;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.OptionalInt;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+
+import static com.facebook.presto.execution.buffer.PagesSerdeUtil.readSerializedPages;
+import static com.facebook.presto.execution.buffer.PagesSerdeUtil.writeSerializedPage;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.Objects.requireNonNull;
+
+public class PrestoSparkTaskCompiler
+        implements IPrestoSparkTaskCompiler
+{
+    private static final Logger log = Logger.get(PrestoSparkTaskCompiler.class);
+
+    private final SessionPropertyManager sessionPropertyManager;
+
+    private final JsonCodec<SparkTaskDescriptor> sparkTaskDescriptorJsonCodec;
+    private final JsonCodec<TaskStats> taskStatsJsonCodec;
+
+    private final Executor notificationExecutor;
+    private final ScheduledExecutorService yieldExecutor;
+
+    private final LocalExecutionPlanner localExecutionPlanner;
+    private final BlockEncodingSerde blockEncodingSerde;
+
+    private final DataSize maxUserMemory;
+    private final DataSize maxTotalMemory;
+    private final DataSize maxSpillMemory;
+
+    private final boolean perOperatorCpuTimerEnabled;
+    private final boolean cpuTimerEnabled;
+
+    @Inject
+    public PrestoSparkTaskCompiler(
+            SessionPropertyManager sessionPropertyManager,
+            JsonCodec<SparkTaskDescriptor> sparkTaskDescriptorJsonCodec,
+            JsonCodec<TaskStats> taskStatsJsonCodec,
+            Executor notificationExecutor,
+            ScheduledExecutorService yieldExecutor,
+            LocalExecutionPlanner localExecutionPlanner,
+            BlockEncodingSerde blockEncodingSerde,
+            TaskManagerConfig taskManagerConfig,
+            NodeMemoryConfig nodeMemoryConfig,
+            NodeSpillConfig nodeSpillConfig)
+    {
+        this(
+                sessionPropertyManager,
+                sparkTaskDescriptorJsonCodec,
+                taskStatsJsonCodec,
+                notificationExecutor,
+                yieldExecutor,
+                localExecutionPlanner,
+                blockEncodingSerde,
+                requireNonNull(nodeMemoryConfig, "nodeMemoryConfig is null").getMaxQueryMemoryPerNode(),
```

Comment:
I am thinking in the future, we should have specific Presto-on-Spark memory configuration, since the Presto node memory config might not make sense in some case (now each "Node" is a Spark Executor).

By default, `MaxQueryMemoryPerNode` is 10% of heap memory, and `MaxQueryTotalMemoryPerNode` is 30% of heap memory.  For Presto-on-Spark, we only need to memory limit (i.e. total memory limit per executor). And reserved pool doesn't make sense for Presto-on-Spark since there is only one query in Spark Task/Executor)

I don't think we need to fix it for the initial commit. But let's add a TODO so we don't forgot it :) 


https://github.com/prestodb/presto/blob/6436aa6ca8663cd31e2b1f7ffcd6670ecacef02a/presto-main/src/main/java/com/facebook/presto/memory/NodeMemoryConfig.java#L33-L36

Reply:
I think this has to be configured with the properties file.
As for the `NodeMemoryConfig` defaults - I'm not sure if it makes a lot of sense to change them. At least for now. Later on we will have a bigger question of extracting the evaluation part into a separate module, and abstracting it out from the conventional Presto deployment model. But for now, let's leave it as is.

Reply:
@arhimondr : Yeah . I found just set `query.max-memory-per-node=X.XGB` works well now . So no need to sweat for now.